core/modules/loader/DoList.h

```diff
@@ -0,0 +1,195 @@
+        std::lock_guard<std::mutex> lock(_mtx);
+        if (_command == nullptr) {
+            if (_isOneShotDone()) { return nullptr; }
+            if ((_needInfo || _timeOut.due(now)) && _timeRequest.due(now)) {
+                _timeRequest.triggered();
+                    _command = createCommand();
+                return _command;
+            }
+        } else if (_command->isFinished()) {
+            _command.reset(); // Allow the command to be sent again later.
+        }
+        return nullptr;
+    }
+
+    bool isAlreadyOnList() { return _addedToList; }
+
+    /// Returns original value of _addedToList.
+    bool setAddedToList(bool value) {
+        return _addedToList.exchange(value);
+    }
+
+    bool removeFromList() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        return (_isOneShotDone() || _remove);
+    }
+
+    /// The info has been updated, so no need to ask for it for a while.
+    void infoReceived() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = false;
+        _timeOut.triggered();
+    }
+
+    void setNeedInfo() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = true;
+    }
+
+    DoListItem::Ptr getDoListItemPtr() {
+        return shared_from_this();
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut.setTimeOut(timeOut); }
+
+    virtual util::CommandTracked::Ptr createCommand()=0;
+
+protected:
+    std::atomic<bool>    _addedToList{false}; ///< True when added to a DoList
+
+    bool    _oneShot{false}; ///< True if after the needed information is gathered, this item can be dropped.
+    bool    _needInfo{true}; ///< True if information is needed.
+    bool    _remove{false}; ///< set to true if this item should no longer be checked.
+    TimeOut _timeOut{std::chrono::minutes(5)}; ///< If no info is needed, check for info after this period of time.
+    TimeOut _timeRequest{std::chrono::seconds(5)}; ///< Rate limiter, no more than 1 message every 5 seconds
+    util::CommandTracked::Ptr _command;
+    std::mutex _mtx; ///< protects _timeOut, _timeRequest, _command, _oneShot, _needInfo
+
+private:
+    /// Lock _mtx before calling.
+    bool _isOneShotDone() {
+        return (!_needInfo && _oneShot);
+    }
+};
+
+
+/// A list of things that need to be done with timers.
+/// Everything on the list is checked, if it's timer has expired, it is queued and the timer reset.
+/// If it is a single use item, it is deleted after completion.
+class DoList {
+public:
+    DoList(Central& central) : _central(central) {}
+    DoList() = delete;
+    DoList(DoList const&) = delete;
+    DoList& operator=(DoList const&) = delete;
+
+    ~DoList() = default;
+
+    void checkList();
+    bool addItem(DoListItem::Ptr const& item) {
+        if (item == nullptr) return false;
+        if (item->isAlreadyOnList()) return false; // fast atomic test
+        {
+            std::lock_guard<std::mutex> lock(_addListMtx);
```

Comment:
The way this method locks the mutex contradicts to the comments to the data members. My understanding is that `_listMtx` should be always locked first like it's done by method `DoList::checkList()`

Reply:
If you have to lock both of them, you must lock _list first. If you know only one will be locked, you only need to lock one of them.